go/internal/tinder/driver_multi.go

```diff
@@ -48,126 +84,189 @@ func (md *MultiDriver) Advertise(ctx context.Context, ns string, opts ...p2p_dis
-				zap.String("key", ns),
-				zap.Error(err))
-
-			continue
-		}
+	cancelSearch()
+	return nil, multierr.Combine(errReturn...)
+}
 
-		driverRefs = append(driverRefs, driver.Name())
-		selCases = append(selCases, reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(ch),
-		})
+// FindPeersAsync expect the caller to use it asynchronously and cancel the context with enough peers or found (or deadline exceeded).
+// If the context is not canceled the search will go indefinitely.
+func (md *MultiDriver) FindPeersAsync(ctx context.Context, outPeers chan<- p2p_peer.AddrInfo, ns string, opts ...p2p_discovery.Option) error {
+	var options p2p_discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return err
 	}
-
-	ndrivers := len(selCases) - 1 // we dont want to wait for the context
-	if ndrivers == 0 {
-		md.logger.Error("no drivers available to find peers")
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
 	}
 
-	cpeers := make(chan p2p_peer.AddrInfo, ndrivers)
-	go func() {
-		defer cancel()
-		defer close(cpeers)
-
-		for ndrivers > 0 {
-			idx, value, ok := reflect.Select(selCases)
-
-			// context has been cancel stop and close chan
-			if idx == selDone {
-				md.logger.Debug("find peers done", zap.Error(ctx.Err()))
-				return
-			}
-
-			if !ok {
-				// The chosen channel has been closed, so zero out the channel to disable the case
-				selCases[idx].Chan = reflect.ValueOf(nil)
-				ndrivers--
-				continue
-			}
-
-			// we can safly get our peer
-			peer := value.Interface().(p2p_peer.AddrInfo)
-			// fmt.Printf("[multi] found a peers: %v for %s\n", peer, ns)
+	// Buffer to try to limit context switch while returning peers.
+	inPeers := make(chan p2p_peer.AddrInfo, 10)
 
-			md.logger.Debug("found a peer",
-				zap.String("driver", driverRefs[idx]),
-				zap.String("key", ns),
-				zap.String("peer", peer.ID.String()))
-
-			// forward the peer
-			cpeers <- peer
+	// Start the fetchers first.
+	ctx, cancelSearch := context.WithCancel(ctx)
+	ndrivers := len(md.drivers)
+	errReturn := make([]error, ndrivers)
+	for i := 0; i < ndrivers; i++ {
+		errReturn[i] = md.drivers[i].FindPeersAsync(ctx, inPeers, ns, opts...)
+	}
+	for _, v := range errReturn {
+		if v == nil {
+			// At least one is working
+			go func() {
+				// We could maybe mux multiple FindPeers with the same key but that will add an other syncing cost.
+				var alreadySeen []p2p_peer.ID
+				var peerCount int
+			MainEmptyingLoop:
+				for peerCount < limit {
+					select {
+					case info := <-inPeers:
+						id := info.ID
+						for _, w := range alreadySeen {
+							if w == id {
+								continue MainEmptyingLoop
+							}
+						}
+						alreadySeen = append(alreadySeen, id)
+						select {
+						case outPeers <- info:
+						case <-ctx.Done():
+							return
+						}
+					case <-ctx.Done():
+						return
+					}
+					peerCount++
+				}
+				cancelSearch()
+			}()
```

Comment:
missing close channel inPeers

Reply:
I don't really care, the context is cancelled and all goroutines having a reference to `inPeers` are about to die.
Also that would add an other waitGroup to avoid `panic: send on closed channel`, I prefer to let the GC take care of that.

Reply:
yes, please discard this review, I mistake inPeers or outPeers.